src/ctap/storage.rs

```diff
@@ -892,4 +994,68 @@ mod test {
         );
         assert_eq!(persistent_store.aaguid().unwrap(), key_material::AAGUID);
     }
+
+    #[cfg(feature = "with_ctap2_1")]
+    #[test]
+    fn test_min_pin_length() {
+        let mut rng = ThreadRng256 {};
+        let mut persistent_store = PersistentStore::new(&mut rng);
+
+        // The minimum PIN lenght is initially at the default.
+        assert_eq!(persistent_store.min_pin_length(), DEFAULT_MIN_PIN_LENGTH);
+
+        // Changes by the setter are reflected by the getter..
+        let new_min_pin_length = 8;
+        persistent_store.set_min_pin_length(new_min_pin_length);
+        assert_eq!(persistent_store.min_pin_length(), new_min_pin_length);
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    #[test]
+    fn test_min_pin_length_rp_ids() {
+        let mut rng = ThreadRng256 {};
+        let mut persistent_store = PersistentStore::new(&mut rng);
+
+        // The minimum PIN lenght is initially at the default.
+        assert_eq!(
+            persistent_store._min_pin_length_rp_ids(),
+            _DEFAULT_MIN_PIN_LENGTH_RP_IDS
+        );
+
+        // Changes by the setter are reflected by the getter..
+        let rp_ids = vec![String::from("example.com")];
+        assert_eq!(
+            persistent_store._set_min_pin_length_rp_ids(rp_ids.clone()),
+            Ok(())
+        );
+        assert_eq!(persistent_store._min_pin_length_rp_ids(), rp_ids);
```

Comment:
The expected value is not really `rp_ids` but `merge(_DEFAULT, rp_ids)`. It doesn't matter for now because `_DEFAULT` is empty. But if we think the default won't be empty, we should probably write the correct expectation.

Reply:
Fixed. How do you think should be approach `MAX_RP_IDS_LENGTH`? The specification simply says: If you can't store the list of RP IDs, return `CTAP2_ERR_KEY_STORE_FULL`. And I believe storage v2 will have that feature? I thought we probably should have another mechanism to block too large requests in v1 and v2, just to have at least some guarantees for how many credentials we can store.

Reply:
Both v1 and v2 reject too large entries. However only v1 returns "StoreFull" error, v2 returns "InvalidArgument" since the error is on the client. Both v1 and v2 reject with "StoreFull" when there is not enough capacity.

The `MAX_RP_IDS_LENGTH` is CTAP-specific and unrelated to the persistent store. The logic should be implemented here in the CTAP persistent storage. Regarding provisioning, it also has to be done in CTAP since it is CTAP-specific. For v2, the way I did it is to limit the number of instances per object type and account for the maximum size if all instances are used with maximal size. [Here](https://github.com/ia0/OpenSK/blob/d41a8fc10a9bfc1697d1f5674c31f1e54746d76a/src/ctap/storage/key.rs#L54-L87) is the mapping from keys to objects (which defines the maximum number of instances per object type). To compute the minimum required capacity, one has to know the maximum size for each object type.

Reply:
Let's chat offline and resolve this in the extension that is actually using `MAX_RP_IDS_LENGTH`.